stdlib/lazy.ml

```diff
@@ -78,3 +78,64 @@ let map_val f x =
   if is_val x
   then from_val (f (force x))
   else lazy (f (force x))
+
+
+
+module Atomic_repeating = struct
+  (* we define these as primitives to avoid a dependency on Printexc *)
+  type raw_backtrace
+  external get_raw_backtrace:
+    unit -> raw_backtrace = "caml_get_exception_raw_backtrace"
+  external raise_with_backtrace: exn -> raw_backtrace -> 'a
+    = "%raise_with_backtrace"
+
+  type 'a ops = {
+    make : unit -> 'a;
+    discard : 'a -> unit;
+  }
+
+  type 'a state =
+    | Thunk of 'a ops
+    | Forcing of 'a ops
+    | Val of 'a
+    | Failed of exn * raw_backtrace
+
+  type 'a t = 'a state Atomic.t
+
+  let from_val v = Atomic.make (Val v)
+  let from_fun ?(discard = ignore) f =
+    Atomic.make (Thunk { make = f; discard })
+
+  let finish ops =
+    match ops.make () with
+    | exception exn ->
+        let bt = get_raw_backtrace () in
+        Failed (exn, bt)
+    | v ->
+        Val v
+
+  let rec force th =
+    match Atomic.get th with
+    | Val v -> v
+    | Failed (exn, bt) ->
+      raise_with_backtrace exn bt
+    | (Thunk ops) as thunk ->
+      ignore (Atomic.compare_and_set th thunk (Forcing ops));
+      (* [compare_and_set] can fail when another domain
+         has set the thunk to [Forcing] or a finished state. *)
+      force th
+    | (Forcing ops) as forcing ->
+        let finished = finish ops in
+        (* [compare_and_set] can fail when another domain
+           has set the thunk to a finished state. In this
+           case our [finished] value is discarded. *)
+        if not (Atomic.compare_and_set th forcing finished)
+        then begin match finished with
+          | Val v ->
+              (* Ignore exceptions raised by discard: we already
+                 have a finished result to return. *)
+              (try ops.discard v with _ -> ())
+          | Thunk _ | Forcing _ | Failed _ -> ()
```

Comment:
If I understand correctly, `force` will ignore the exceptions that are raised by `ops.make` of the loser in the race. 

The `force` doc string is sufficiently ambiguous now to capture this?

> If there is a race between several calls to [force], the computation may be repeated and some of the finished **_results_** will be discarded.

(emphasis mine)

"results" can be considered to include both values and exceptions?

Reply:
Right, I used "result" to mean "value returned or exception raised", but I agree that being explicit is better. Here is the new `force` documentation string. The first paragraph is reused from `Lazy.force`:

```
  (** [force x] forces the suspension [x]. If [x] has already been
      forced, [Lazy.force x] returns the same value again without
      recomputing it. If it raised an exception, the same exception is
      raised again.

      If there is a race between several calls to [force], the
      computation may be repeated several times. But all calls will
      return the same value or raise the same exception; the other
      values or exceptions will be discarded. *)
```